runtime/src/anchor.rs

```diff
@@ -0,0 +1,142 @@
+//! Minimal proof of existence registry.
+//!
+//! Anchors are hashed once before being added to storage. To check whether an anchor exists
+//! query the "Anchors" map for the hash of the anchor. If a corresponding value exists, then the
+//! anchor exists and the value represents the block number when it was first published.
+
+use alloc::vec::Vec;
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage, dispatch::DispatchResult, ensure,
+    traits::Get,
+};
+use frame_system::{self as system, ensure_signed};
+use sp_runtime::traits::Hash;
+
+pub trait Trait: system::Trait {
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        /// The anchor being posted was already created in a previous block.
+        AnchorExists,
+    }
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Blob {
+        // Hasher can be the identity here becuse we perform a hash ourself which has the same
+        // merkle-trie balancing effect as using a hash-prefix map.
+        Anchors: map hasher(identity) <T as system::Trait>::Hash =>
+            Option<<T as system::Trait>::BlockNumber>;
+    }
+}
+
+decl_event! {
+    pub enum Event<T>
+    where
+        AccountId = <T as system::Trait>::AccountId,
+        BlockNumber = <T as system::Trait>::BlockNumber,
+        Hash = <T as system::Trait>::Hash,
+    {
+        /// A new permanent anchor was posted.
+        AnchorDeployed(Hash, AccountId, BlockNumber),
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        fn deposit_event() = default;
+
+        /// Drop a permanent anchor.
+        #[weight = T::DbWeight::get().reads_writes(1, 1)]
+        pub fn deploy(
+            origin,
+            dat: Vec<u8>,
+        ) -> DispatchResult {
+            Module::<T>::deploy_(origin, dat)
+        }
+    }
+}
+
+impl<T: Trait> Module<T> {
+    fn deploy_(origin: <T as system::Trait>::Origin, dat: Vec<u8>) -> DispatchResult {
+        let acct = ensure_signed(origin)?;
+
+        // check
+        let h = <T as system::Trait>::Hashing::hash(&dat);
+        ensure!(Anchors::<T>::get(&h).is_none(), Error::<T>::AnchorExists);
+
+        // execute
+        let bn = <system::Module<T>>::block_number();
+        Anchors::<T>::insert(&h, &bn);
+        Self::deposit_event(Event::<T>::AnchorDeployed(h, acct, bn));
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
These tests pretty much just restate what the implementation does. The module is simple enough that there really isn't much to test.